### DISCUSSION TITLE
[Question] New recent activity and export features

### DISCUSSION THREAD

**Discussion by scmanjarrez** [acheong08/EdgeGPT#432] [General]
With latest version of Bing Chat, it's possible to retrieve old chats. Do you know how it works? I still get expiration date for conversations (6h), but it seems that if you use Edge the chat are kept indefinitely.

Also, have you checked if it's possible to retrieve the link to download all messages in the conversation?

**Comment by acheong08** [comment]
I haven't reversed yet.

**Reply by acheong08** [reply]
On a to do list but my time is a bit limited

**Comment by scmanjarrez** [comment]
I've seen you added the ability to retrieve conversations from a file, however I've seen in the requests done by Edge that you can HTTP GET this endpoint to receive the list of current open chats:
```
https://www.bing.com/turing/conversation/chats
```
And you get a json with the format:
```
(root) {…}
chats [9]
result {…}
value Success
message 9 chats found.
serviceVersion 20230607.30
clientId xxxxxxxxxxxxx
```
chats:
```
{
            "conversationId": "51D|BingProd|FB1E297BB177829F519A8D833EBD9CF90477A2C3EF53FDACF2A33E9D310AA9B7",
            "chatName": "bing ai chat",
            "conversationSignature": "QADke0k7lLRsJTnPMkbxG6vgAU1Bp3yxnkuX/sQGHZo=",
            "tone": "Creative",
            "createTimeUtc": 1685787858222,
            "updateTimeUtc": 1685787861617
        },
        {
            "conversationId": "51D|BingProd|641EF5F3973DAEC6DE95A9FB59C3EB5043C0D04BA71DCEB66F48960505CA945C",
            "chatName": "bing ai chat",
            "conversationSignature": "YhJ12nX5F28rTmET6iXJX1zjrO4810uO\u002BGWOnlvjIIs=",
            "tone": "Creative",
            "createTimeUtc": 1685143716103,
            "updateTimeUtc": 1685143719101
        },
```

Isn't possible to retrieve a conversation just by the conversationId and conversationSignature? You can infer the client_id and invocation_id

**Reply by acheong08** [reply]
Yes

**Reply by samuelcaldas** [reply]
how?

**Reply by scmanjarrez** [reply]
I'll make a PR, check this https://github.com/scmanjarrez/EdgeGPT-Telegram-Bot/blob/master/src/utils.py#L129

**Reply by scmanjarrez** [reply]
Basically you need to send a GET to https://www.bing.com/turing/conversation/chats, setting the header `{"Cookie": "SUID=A; _U=<value_of_cookie>;"}`. Then you use the client_id returned in the request and use invocation_id 4 (I haven't tested bigger values, but it should be >3)

**Comment by scmanjarrez** [comment]
Related #532
